tuf/client_rework/metadata_bundle.py

```diff
@@ -0,0 +1,505 @@
+# Copyright the TUF contributors
+# SPDX-License-Identifier: MIT OR Apache-2.0
+
+"""TUF client bundle-of-metadata
+
+MetadataBundle keeps track of current valid set of metadata for the client,
+and handles almost every step of the "Detailed client workflow" in the TUF
+specification (the remaining steps are download related). The bundle takes
+care of persisting valid metadata on disk, loading local metadata from disk
+and deleting invalid local metadata.
+
+Loaded metadata can be accessed via the index access with rolename as key
+or, in the case of top-level metadata using the helper properties like
+'MetadataBundle.root'
+
+The rules for top-level metadata are
+ * Metadata is loadable only if metadata it depends on is loaded
+ * Metadata is immutable if any metadata depending on it has been loaded
+ * Caller must load/update these in order:
+   root -> timestamp -> snapshot -> targets -> (other delegated targets)
+ * Caller should try loading local file before updating metadata from remote
```

Comment:
Root is handled separately: local root load happens during instance initialization. I make that clear in this text.

For all others I wanted to leave the decision to caller: it would be possible to keep track of local loads within the bundle and to do it automatically when needed, but it's more code and the caller is going to do it anyway (because in several cases caller wants to know if the local metadata is valid before it decides whether to even try downloading remote metadata)...